### PR TITLE
FiveCrop added to preprocessing

### PR DESCRIPTION
**Pull request by sabribarac**
# What does this PR do?

This PR adds FiveCrop ported over from TorchVision. Crops image into 5 crops, 
top_left, top_right, bottom_left, bottom_right and center crop.

## Before submitting
- [x] This PR adds FiveCrop ported over from TorchVision
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Did you write any new necessary tests?

## Who can review?

@LukeWood @tanzhenyu @ianstenbit

### CODE REVIEW

File: keras_cv/layers/__init__.py [hunk 1]
```diff
@@ -117,3 +117,4 @@
 from keras_cv.layers.spatial_pyramid import SpatialPyramidPooling
 from keras_cv.layers.transformer_encoder import TransformerEncoder
 from keras_cv.layers.vit_layers import PatchingAndEmbedding
+from keras_cv.layers.preprocessing.five_crop import FiveCrop
```
Comment (ianstenbit) [inline keras_cv/layers/__init__.py:120]: nit: please keep imports sorted alphabetically in this file

File: keras_cv/layers/preprocessing/__init__.py [hunk 2]
```diff
@@ -79,3 +79,4 @@
 from keras_cv.layers.preprocessing.vectorized_base_image_augmentation_layer import (
     VectorizedBaseImageAugmentationLayer,
 )
+from keras_cv.layers.preprocessing.five_crop import FiveCrop
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/__init__.py:82]: nit: please keep imports sorted alphabetically in this file

File: keras_cv/layers/preprocessing/five_crop.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            raise ValueError(msg.format(self.size, (image_height, image_width)))
+
+        tl = tf.image.crop_to_bounding_box(image, 0, 0, crop_height, crop_width)
+        tr = tf.image.crop_to_bounding_box(image, 0, image_width-crop_width, crop_height, crop_width)
+        bl = tf.image.crop_to_bounding_box(image, image_height - crop_height, 0, crop_height, crop_width)
+        br = tf.image.crop_to_bounding_box(image, image_height - crop_height, image_width - crop_width, crop_height, crop_width)
+
+        center = tf.keras.layers.CenterCrop(crop_height, crop_width)(image)
+
+        return tl, tr, bl, br, center
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/five_crop.py:79]: I think returning a tuple here is probably not the right API surface. 

I would guess we want to concatenate these along the batch axis. @LukeWood is our preprocessing pro -- what do you think should be the approach here?

File: keras_cv/layers/preprocessing/five_crop.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    BaseImageAugmentationLayer,
+)
+
+
+@keras.utils.register_keras_serializable(package="keras_cv")
+class FiveCrop(BaseImageAugmentationLayer):
+    """ Crops image into four courners and one center crop.
+
+    Args:
+        size: a tuple or a int. Desired output size to crop.
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/five_crop.py:28]: Let's use `height` and `width` as separate parameters, to be consistent with [`RandomCrop`](https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/random_crop.py) (we can also model the docstrings on that layer's docstrings).

File: keras_cv/layers/preprocessing/five_crop.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+from tensorflow import keras
+
+from keras_cv.layers.preprocessing.base_image_augmentation_layer import (
+    BaseImageAugmentationLayer,
+)
+
+
+@keras.utils.register_keras_serializable(package="keras_cv")
+class FiveCrop(BaseImageAugmentationLayer):
+    """ Crops image into four courners and one center crop.
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/five_crop.py:25]: Is there a paper or other documentation we can reference somewhere about the value of this preprocessing approach?

If so, it would be great to add a pointer to it in the docstring.

File: keras_cv/layers/preprocessing/five_crop.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        br = tf.image.crop_to_bounding_box(image, image_height - crop_height, image_width - crop_width, crop_height, crop_width)
+
+        center = tf.keras.layers.CenterCrop(crop_height, crop_width)(image)
+
+        return tl, tr, bl, br, center
+
+    def augment_labels(self, labels, transformations, **kwargs):
+        return labels
+
+    def augment_segmentation_masks(
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/five_crop.py:84]: This no-op transformation of segmentation masks is inconsistent with the changes to the input image.

To support augmenting the segmentation masks, we should apply the same transformation here that we do to the images (doing a 5-crop and returning the 5 masks separately)

Comment (ianstenbit) [reply keras_cv/layers/preprocessing/five_crop.py:84]: (The same is true for the `bounding_boxes`)

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/layers/preprocessing/five_crop.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/five_crop.py:79]: I was about to comment the same thing :)

**Comment by LukeWood** [comment]
Thanks for your enthusiasm in porting this!

Would you be able to open a GitHub issue covering what this technique is used for?  That way we can discuss where (and if) it fits into the library.

Thanks again for your enthusiasm!

**Comment by innat** [comment]
@LukeWood 

> covering what this technique is used for

One of an issue case is enabling learning feature correspondence in terms of self-supervision. For example, [STEGO](https://github.com/mhamilton723/STEGO).

@sabribarac 
If you plan to move forward, please follow luke's suggestion (open a ticket first).

**Comment by sabribarac** [comment]
> @LukeWood
> 
> > covering what this technique is used for
> 
> One of an issue case is enabling learning feature correspondence in terms of self-supervision. For example, [STEGO](https://github.com/mhamilton723/STEGO).
> 
> @sabribarac If you plan to move forward, please follow luke's suggestion (open a ticket first).

Thanks @innat, will do
